core/src/solution_submission.rs

```diff
@@ -215,29 +222,60 @@ async fn retry_with_gas_price_increase(
                 solution.clone(),
                 claimed_objective_value,
                 gas_price,
-                if gas_price == gas_cap {
-                    None
-                } else {
-                    Some(BLOCK_TIMEOUT)
-                },
+                Some(BLOCK_TIMEOUT),
             )
             .await;
 
-        let gas_price_can_still_increase =
-            gas_price.as_u128() as f64 * MIN_GAS_PRICE_INCREASE_FACTOR < gas_cap.as_u128() as f64;
+        let before_deadline = deadline
+            .map(|deadline| Instant::now() < deadline)
+            .unwrap_or(true);
+        let gas_price_can_still_increase = gas_price.as_u128() as f64
+            * MIN_GAS_PRICE_INCREASE_FACTOR
+            < gas_price_cap.as_u128() as f64;
+
         // Breaking condition for our loop
-        gas_price_can_still_increase
-            && matches!(
-                result,
-                Err(MethodError {
-                    inner: ExecutionError::ConfirmTimeout,
-                    ..
-                })
-            )
+        match result {
+            Err(MethodError {
+                inner: ExecutionError::ConfirmTimeout,
+                ..
+            }) => {
+                if gas_price_can_still_increase && before_deadline {
+                    true
+                } else {
+                    let gas_price = if gas_price_can_still_increase {
+                        gas_price_cap
+                    } else {
+                        U256::from(
+                            (gas_price.as_u128() as f64 * MIN_GAS_PRICE_INCREASE_FACTOR) as u128,
+                        )
+                    };
+                    log::info!(
+                        "trying to cancel solution submission transaction for batch {}",
+                        batch_index
+                    );
+                    // `contract.noop_transaction` gets the current nonce from the node which is
+                    // technically a race condition. It would be better to retrieve the nonce from
+                    // the `submit_solution` transaction but I couldn't find a way to do that.
```

Comment:
This can be done a couple of ways:
1. Retrieve the transaction data with the transaction hash: https://docs.rs/web3/0.11.0/web3/api/struct.Eth.html#method.transaction. Unfortunately, the transaction hash is not super easy to determine right now and requires first `TransactionBuilder::build`-ing the transaction and using `web3` to send it. I created gnosis/ethcontract-rs#345 to address the issue.
2. Manually setting the nonce for the submit solution transaction by as you do in the `noop_transaction`, and then returning along side the future.

Reply:
Let's keep it this way until the ethcontract PR is available.

Reply:
The same race condition applies for increasing the gas price for solution submission.

Reply:
In general though, I think it would be better to ensure the same nonce that was used by the `submit_solution` transaction is used for the cancellation transaction.